### ISSUE TITLE
MSTP did not work on the DSA felix switch

### ISSUE THREAD

**Issue by shockingchen** [mstpd/mstpd#129]
![image](https://user-images.githubusercontent.com/49074914/138416799-2471c3fa-24cc-47cc-9431-2a2301943f93.png)
when I config the mstp on 2 switches,add one mstp tree with id 1 and vlan 100,but it dont work.here is my config:
1/1:eno0    1/2:swp0       1/3:swp1     1/4:swp2       1/5:swp3
root@forlinx:~# bridge vlan
port              vlan-id  
eno0              1 PVID Egress Untagged
swp0              1 PVID Egress Untagged
swp1              1 PVID Egress Untagged
swp2              100
                  200
swp3              100
                  200
Bridge            1 PVID Egress Untagged
                  100
                  200
docker0           1 PVID Egress Untagged

root@forlinx:~# mstpctl showmstilist Bridge
Bridge list of known MSTIs:
 0 1

root@forlinx:~# mstpctl showfid2mstid Bridge
Bridge FID-to-MSTID allocation table:
  MSTID 0: 0,2-4095
  MSTID 1: 1
root@forlinx:~# mstpctl showvid2fid Bridge             
Bridge VID-to-FID allocation table:
  FID 0: 1-99,101-4094
  FID 1: 100
 
 
 root@forlinx:~# mstpctl showtree Bridge 1
Bridge MSTI 1 info
  bridge id          1.001.2E:95:57:EC:2C:9C
  regional root      1.001.2E:95:57:EC:2C:9C
  root port          none
  internal path cost 0
  time since topology change 1407
  topology change count      4
  topology change            no
  topology change port       swp2
  last topology change port  swp3
  
  
 root@forlinx:~# mstpctl showtreeport Bridge swp2 1 
Bridge:swp2 MSTI 1 info
  role               Master                  port id              8.004
  state              forwarding              disputed             no
  internal port cost 20000                   admin internal cost  0
  dsgn regional root 1.001.2E:95:57:EC:2C:9C dsgn internal cost   0
  designated bridge  1.001.2E:95:57:EC:2C:9C designated port      8.004
  
  
  root@forlinx:~# mstpctl showtreeport Bridge swp3 1  
Bridge:swp3 MSTI 1 info
  role               Alternate               port id              8.005
  state              discarding              disputed             no
  internal port cost 20000                   admin internal cost  0
  dsgn regional root 1.001.2E:95:57:EC:2C:9C dsgn internal cost   0
  designated bridge  1.001.2E:95:57:EC:2C:9C designated port      8.005

then I change the swp2's  internal port cost to 2000000,but it still seems Master role,the MSTP did not work on the vlan 100.
 root@forlinx:~# mstpctl showtreeport Bridge swp2 1 
Bridge:swp2 MSTI 1 info
  role               Master                  port id              8.004
  state              forwarding              disputed             no
  internal port cost 2000000                   admin internal cost  0
  dsgn regional root 1.001.2E:95:57:EC:2C:9C dsgn internal cost   0
  designated bridge  1.001.2E:95:57:EC:2C:9C designated port      8.004

how to config the MSTP on the switch?or is it support MSTP on it?

thanks

**Comment by pavlix** [comment]
> https://lwn.net/Articles/884996/ is that something MSTPD needs?

Yes. This is necessary to connect mstpd with hardware that supports MST instances. We experimented with a custom patchset before this was ready.

**Comment by KanjiMonster** [comment]
I created a proof of concept implementation using that MST support. I tested it with local changes adding MST support to b53 (where most of the chips support it).
